### PR TITLE
[Security] Support RSA algorithm signature for OIDC tokens

### PR DESCRIPTION
**Pull request by louismariegaborit**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | 
| License       | MIT

Add support for RSA signature algorithm for OidcTokenHandler.  
Amazon Cognito uses RS256 algorithm for its tokens.

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php]:
```suggestion
                throw new InvalidArgumentException(sprintf('Unsupported signature algorithm "%s". Only ES* and RS* algorithms are supported. If you want to use another algorithm, create your TokenHandler as a service.', $algorithm));
```

File: src/Symfony/Component/Security/Http/composer.json
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/composer.json]:
```suggestion
        "web-token/jwt-signature-algorithm-rsa": "^3.1"
```

**Review by nicolas-grekas — APPROVED** [review]

**Comment by Spomky** [comment]
Hi,

Is it possible to put this PR on hold?
The reason is that algorithms from the `web-token/` suite will be grouped and, very soon, with `web-token/jwt-library` you will have all at once (including encryption to help #50441)

**Comment by Spomky** [comment]
Hi @louismariegaborit,

Can you test with `"web-token/jwt-library": "3.3.0@dev"` adn `"ext-openssl": "*"`?
Both ` web-token/jwt-signature-algorithm-ecdsa` and `web-token/jwt-signature-algorithm-rsa` dependencies can be removed.

Also, because the new `web-token/jwt-library` contains more than `ESxxx` and `RSxxx` algorithms, I am wondering if this PR could be renamed to `[Security] Support any signature algorithms for OIDC tokens` and allow developers picking the algorthims they need.

Note: `web-token/jwt-library` will be tagged in a couple of days.

**Comment by louismariegaborit** [comment]
Hi @Spomky 

This seems ok. I removed also `web-token/jwt-checker`.
For the `ext-openssl`, I added it in the require-dev but it's in suggest like in your repository that you want that be added ?

I will look so that developers picked the algorithms they need.

I will update the PR title when we validate the work.

**Comment by louismariegaborit** [comment]
@Spomky I did a try. WDYT ?
I don't know if I can delete old algorithm services without deprecations. And if I keep the services with deprecations, could you tell me how do it ?

**Comment by Spomky** [comment]
> This seems ok. I removed also web-token/jwt-checker.
> For the ext-openssl, I added it in the require-dev but it's in suggest like in your repository that you want that be added ?

Indeed, most of `jwt-*` packages are now under `jwt-library`.
OK for `ext-openssl` in the `required-dev` section. I'll make sure the algorithms complain if it's missing.

---

@vincentchalamon do you have any recommendation to have a better algorithm support architecture.
What could be interesting is to allow multiple algorithms

```yaml
# config/packages/security.yaml
security:
    firewalls:
        main:
            access_token:
                token_handler:
                    oidc:
                        # Algorithms used to sign the JWS
                        algorithms:
                            - 'ES256'
                            - 'RS256'
                            - 'PS256'
                        # A JSON-encoded JWK
                        key: '{"kty":"...","k":"..."}'
```

From my understanding, it will require:
* If the configuraition key `algorithm` is present: to normaliwe the current configuration fro; `["algorithm" => $alg]` to `["algorithms" => [$alg]]`
* Change the `OidcTokenHandler` first argument to accept an algorithm manager object instead a single algorithm
* Use the [algorithm manager factory](https://github.com/web-token/jwt-framework/blob/3.3.x/src/Library/Core/AlgorithmManagerFactory.php#L54) to create the algorithm manager according to the configuration

Any ideas on this?

**Comment by louismariegaborit** [comment]
> @vincentchalamon do you have any recommendation to have a better algorithm support architecture. What could be interesting is to allow multiple algorithms
> 
> ```yaml
> # config/packages/security.yaml
> security:
>     firewalls:
>         main:
>             access_token:
>                 token_handler:
>                     oidc:
>                         # Algorithms used to sign the JWS
>                         algorithms:
>                             - 'ES256'
>                             - 'RS256'
>                             - 'PS256'
>                         # A JSON-encoded JWK
>                         key: '{"kty":"...","k":"..."}'
> ```
> 
> From my understanding, it will require:
> 
> * If the configuraition key `algorithm` is present: to normaliwe the current configuration fro; `["algorithm" => $alg]` to `["algorithms" => [$alg]]`
> * Change the `OidcTokenHandler` first argument to accept an algorithm manager object instead a single algorithm
> * Use the [algorithm manager factory](https://github.com/web-token/jwt-framework/blob/3.3.x/src/Library/Core/AlgorithmManagerFactory.php#L54) to create the algorithm manager according to the configuration
> 
> Any ideas on this?

@Spomky I started this work in another PR (https://github.com/symfony/symfony/pull/51665).  
I still need to support an array for the algorithms.
I think, we must decide what is the responsability of each PR.

**Comment by vincentchalamon** [comment]
@Spomky on an OIDC server (e.g.: Keycloak), is it possible to allow multiple algorithms on a single realm?

If true, multiple algorithms configuration could be interesting, indeed. I'm just wondering if the `key` configuration (JWK) shouldn't be associated to the corresponding algorithm, as I think they're linked (cf. JWA).

If false, I don't think we should allow multiple algorithms configuration as multiple realms are not supported.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php
Comment (vincentchalamon) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php]: It is not recommended to allow all algorithms available on web-token/jwt-* as some of them are experimental, not recommended or deprecated: https://github.com/web-token/jwt-doc/blob/v2.x/the-components/encrypted-tokens-jwe/encryption-algorithms.md.

**Review by vincentchalamon — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php
Comment (louismariegaborit) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php]: Experimental algorithms are not in the web-token/jwt-library packages.  
But, jwk can be imposed by authentication provider. If the provider use experimental or deprecated algorithms why we couldn't use them ?

**Comment by louismariegaborit** [comment]
> @Spomky on an OIDC server (e.g.: Keycloak), is it possible to allow multiple algorithms on a single realm?
> 
> If true, multiple algorithms configuration could be interesting, indeed. I'm just wondering if the `key` configuration (JWK) shouldn't be associated to the corresponding algorithm, as I think they're linked (cf. JWA).
> 
> If false, I don't think we should allow multiple algorithms configuration as multiple realms are not supported.

When we use JWKSet, we can have multiple jwk for multiple algorithms. The `kty` and `alg` parameter in the JWK define the algorithm for each key.

**Comment by Spomky** [comment]
> @Spomky on an OIDC server (e.g.: Keycloak), is it possible to allow multiple algorithms on a single realm?
> 
> If true, multiple algorithms configuration could be interesting, indeed. I'm just wondering if the key configuration (JWK) shouldn't be associated to the corresponding algorithm, as I think they're linked (cf. JWA).
> 
> If false, I don't think we should allow multiple algorithms configuration as multiple realms are not supported.

To be honest, I am not sure how Keycloak uses algorithms. Same for other OIDC servers (AFAIK Auth0 uses either ES256 or RS256).
I read again the [OIDC Discovery](https://openid.net/specs/openid-connect-discovery-1_0.html) specification and the `id_token_signing_alg_values_supported` returns a list of algorithm values. Even if it is not that common, it seems acceptable for identity servers to support other algs than only RS256.

You are right to mention that a single JWK cannot be used for different algs. In this case, a JWKSet is needed. 
Having JWKSet can be a nice addition, not only for supporting several signing algorithms, but also for supporting key rotation feature without service interuption.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php
Comment (Spomky) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php]: Experimental algorithms are in another package, but the main one also contains `none`.
That's why it is important to explicitly configure allowed algorithms and to have the associated key type.

@louismariegaborit can you grant me write access on your repository? I will try to change the way algorithm services are managed and propose an update on that point.

Comment (louismariegaborit) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php]: @Spomky Invitation sent ;)

File: composer.json
Comment (stof) [inline composer.json]: this should be removed. If openssl is needed to run some tests, we should skip those tests when the extension is missing instead of forbidding to install the dev setup of Symfony entirely (maybe the contributor wants to work on a totally different part of the system)

File: composer.json
Comment (stof) [inline composer.json]: please use `^3.3` instead, which will avoid having to change the requirement once tagged (our dev setup runs with `minimum-stability: dev` anyway)

File: src/Symfony/Component/Security/Http/composer.json
Comment (stof) [inline src/Symfony/Component/Security/Http/composer.json]: same here for the dev requirement

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php]: this error message does not match the check that is done.

**Review by stof — CHANGES_REQUESTED** [review]

File: composer.json
Comment (louismariegaborit) [reply composer.json]: `web-token/jwt-library` is a new library which will tagged in few days.
https://github.com/symfony/symfony/pull/53682#issuecomment-1922051301

**Comment by louismariegaborit** [comment]
> You are right to mention that a single JWK cannot be used for different algs. In this case, a JWKSet is needed. Having JWKSet can be a nice addition, not only for supporting several signing algorithms, but also for supporting key rotation feature without service interuption.

@Spomky I suggest that we keep this PR do support of the new `web-token/jwt-library` and do the refactoring on JWKSet/AlgorithmManager in the https://github.com/symfony/symfony/pull/51665.
I am also opening a discussion about the future of the SecurityBundle configuration for the OidcTokenHandler.

**Comment by Spomky** [comment]
Hi,

I added tests to make sure nothing is impacted by the changes.
* Multiple algorithms are supported.
    * They shall be explicitely listed in the token handler configuration file.
    * If you want to use a custom algorithm, it just has to be tagged as `security.access_token_handler.oidc.signature_algorithm` to be used.
* JWK or JWKSet are supported

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (vincentchalamon) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php]: I suggest to create a jwkset here from the specified key: `{"keys":[%s]}`
So the bundle won't set something deprecated (cf. OIDCTokenHandler: "key" is deprecated in favor of "keyset")

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (vincentchalamon) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:77]: IMHO we have 2 DX possibilities:
1. keep `algorithm` and `key` options in case of single algorithm and single key, which are automatically converted to `algorithms` and `keyset` options
2. or deprecate them and handle the BC layer with `beforeNormalization`, and always initialize `algorithms` and `keyset` options internally to don't create deprecated objects (cf. OIDCTokenHandler)

WDYT?

**Comment by louismariegaborit** [comment]
Good job @Spomky but finally you decide to merge my dedicated PR which support JWKSet where @nicolas-grekas asked me to split in two PRs the support of others algorthms and the support of JWKSet.

**Review by vincentchalamon — COMMENTED** [review]
Nice job! I really like this new approach with JWKSet and multiple algorithms supports with web-token/jwt-library :-)
Don't forget to update the CHANGELOGs of the Security component with your changes.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (Spomky) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php]: I take it!

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (Spomky) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:77]: One question came into my mind after that: if at some point JWEs are supported, we will have other algorithms lists and keyset. What if we name `algorithms` into `verification_algorithms` and `keyset` into `verification_keyset`? (or under a `verification` section)
Anyway, option 2 looks better to me.

Comment (GregoireHebert) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:77]: Like @Spomky, to me it feels natural to look for a JWSignature key and JWEncryption key.
I would deprecate (Vincent's second option) and have  

```yaml
signature:
    algorithms: []
    keyset: []
encryption:
    algorithms: []
    keyset: []
```

Comment (vincentchalamon) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:77]: IMO `signature` seems more a adequat name than `verification`.

However, even if I like this idea, I would prefer to keep it MVP. JWE is not implemented yet. I would suggest to split these configurations in the JWE related Pull Request (once it'll be created).

**Comment by Spomky** [comment]
Hi,

I addressed your comments and the configuration file will be as follows:

```yaml
# config/packages/security.yaml
security:
    firewalls:
        main:
            access_token:
                token_handler:
                    oidc:
                        algorithms: # Algorithms used to verify the JWS
                            - 'ES256'
                            - 'RS256'
                            - 'PS256'
                        keyset: '{"keys":[{"kty":"...","k":"..."}]}' # A JSON-encoded JWKSet (acceptable keys)
```

The previous configuration will be normalized and deprecations are triggered.

**Comment by Spomky** [comment]
> Hi,
> 
> Is it possible to put this PR on hold? The reason is that algorithms from the `web-token/` suite will be grouped and, very soon, with `web-token/jwt-library` you will have all at once (including encryption to help #50441)

`web-token/jwt-library` is now released

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php]: let's use short closures everywhere possible (please update the other occurences):
```suggestion
                    ->ifTrue(static function ($v) => !isset($v['algorithm']) && !isset($v['algorithms']))
```

**Review by nicolas-grekas — COMMENTED** [review]
LGTM. OK for the rest of you?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (vincentchalamon) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:42]: Isn't it a BC break to remove this (abstract) service `security.access_token_handler.oidc.jwk`? AFAIK, it's not marked `@internal` nor `@experimental`

Comment (Spomky) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:42]: Indeed. I will make sure this service exists when a JWK is set

Comment (Spomky) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php:42]: The abstract service `security.access_token_handler.oidc.jwk` is back and marked as deprecated

**Review by vincentchalamon — APPROVED** [review]
LGTM.
@Spomky can you please rebase from symfony:7.1 to fix the CI?

**Comment by Spomky** [comment]
> LGTM. @Spomky can you please rebase from symfony:7.1 to fix the CI?

I am working on it. Some tests are failing due to the changes here.

**Comment by Spomky** [comment]
The CI seems to be fixed. Errors are not related to the changes.
fabbot is asking me to apply an empty patch. Not sure what is causing this error.
Any advice for the next step?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php]: (same below)
```suggestion
                    ->ifTrue(static fn ($v) => !isset($v['algorithm']) && !isset($v['algorithms']))
```

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (nicolas-grekas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php]: (same below)
```suggestion
                    ->then(static function ($v) {
```

**Review by nicolas-grekas — COMMENTED** [review]
LGTM with minor CS fixes

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (Spomky) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php]: 👍🏼done

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php
Comment (Spomky) [reply src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/AccessToken/OidcTokenHandlerFactory.php]: Done as well

**Review by chalasr — DISMISSED** [review]

**Review by chalasr — CHANGES_REQUESTED** [review]
Can you look at failing tests?

**Comment by Spomky** [comment]
Hi @chalasr,

It seems to be fine now.
I corrected one of the tests without the `@group legacy`. Other errors are not related to this PR.

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
Thank you @louismariegaborit and @Spomky.

**Pull request merged**
